proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1284 @@
+  have the same required signature (e.g., ``Str``), then you
+  have a single required signature ``Str`` that is the union
+  of these two signatures: signatures are identified only
+  by module name.  To keep these two requirements separate,
+  you would rename one of the requirements to a different name
+  using the ``backpack-includes`` directive::
+
+    library one-string
+        -- One requirement, named Str
+        build-depends: concat-indef, stringutils-indef
+
+    library two-string
+        -- Two requirements, Str and Str2
+        build-depends: concat-indef, stringutils-indef
+        backpack-include:
+            stringutils-indef requires (Str as Str2)
+
+* In addition to the inherited requirements from dependencies,
+  a user can also define a local ``hsig`` to refine the required
+  signature further (i.e., define extra types).
+
+The current implementation of Backpack in GHC has some notable
+user-facing limitations:
+
+1. It is not possible to define a module in a library, and then
+   use it to immediately to instantiate an indefinite library::
+
+        library concat-bytestring-bad
+            build-depends: concat-indef -- has Str requirement
+            exposed-modules: Str, ConcatUser -- can't use these to fill
+
+   Instead, ``Str`` must be pulled out into a separate library
+   of its own (Cabal 2.0 supports multiple libraries in a package,
+   making this less burdensome.)  The reason for this restriction
+   is to simplify implementation of the build system: if this
+   mode of use was allowed, it would be necesary to first build
+   ``Str``, then build ``concat-indef``, and then come back to
+   ``concat-bytestring-bad`` and finish building the rest of the
+   modules.
+
+   Note, however, it is permissible to inherit a signature while also
+   defining a local signature.
+
+2. Mutual recursion is not allowed.  For example, these libraries
+   cannot be instantiated with each other::
+
+        library p
+            signatures: A
+            exposed-modules: B
+        library q
+            signatures: B
+            exposed-modules: A
+
+   Signature merging can also result in mutual recursion; suppose
+   a library has these two signatures::
+
+        signature A where
+        signature B where
+            import A
+
+   and another library has the import swapped: merging these would
+   result in a cycle between ``A`` and ``B``; thus it is not allowed.
+
+   Eventually we do want to support mutual recursion in all these cases
+   (the theory certainly allows for it), but we declared it as out of
+   scope for the initial release of Backpack.
+
+Identifiers
+-----------
+
+In this section, we describe the grammar of identifiers in Backpack.
+These identifiers are used, for example, to determine the unique
+name associated with each entity in Haskell, which in turn determines
+when two types are equal.
+
+::
+
+    ComponentId ::= [A-Za-z0-9-_.]+
```

Comment:
That's what Cabal's parsing code (predates me) says. We could probably restrict it without causing too much gnashing of teeth though.